x/btcstaking/keeper/keeper.go

```diff
@@ -38,7 +37,6 @@ func NewKeeper(
 
 	btclcKeeper types.BTCLightClientKeeper,
 	btccKeeper types.BtcCheckpointKeeper,
-	ckptKeeper types.CheckpointingKeeper,
```

Comment:
We will have a [design](https://github.com/babylonlabs-io/pm/blob/main/adr/adr-024-timestamping-public-randomness.md#decision) for not assigning voting power to an FP if it does not have finalised pub rand at this height. Will this be implemented in a separate PR?

Also this seems to mean that the voting power table update will need to be moved to the finality module, as the voting power dist algorithm will depend on the pub rand KV store which is in finality module

Reply:
> We will have a [design](https://github.com/babylonlabs-io/pm/blob/main/adr/adr-024-timestamping-public-randomness.md#decision) for not assigning voting power to an FP if it does not have finalised pub rand at this height. Will this be implemented in a separate PR?

Exactly. This PR is to kick off the feature and focus on db schema change.

> Also this seems to mean that the voting power table update will need to be moved to the finality module, as the voting power dist algorithm will depend on the pub rand KV store which is in finality module

Would it cause any concerns if we let the btc staking keeper have some expected interface of finality keeper to this end?

Reply:
Hmm ok, should this PR go to a feature branch then? as the PR itself might let a bunch of blocks unfinalised until we change the criteria of having voting power (and thereby definition of activated height)

> Would it cause any concerns if we let the btc staking keeper have some expected interface of finality keeper to this end?

I think if we do this, then BTC staking module is not repsonsible for inferring voting power of FPs anymore. Rather its sole functionality is to manage state FPs / BTC delegations. Voting power table then turns to be like a "view" of finality module over the BTC staking module.

There is an ancient discussion on this https://babylonlabs.atlassian.net/wiki/spaces/BABYLON/pages/5878176/ADR-018+BTC+staking+consumer+module+refactoring?focusedCommentId=5878896 and it seems that the timestamping pub rand feature mandates this refactor

Reply:
> Hmm ok, should this PR go to a feature branch then? as the PR itself might let a bunch of blocks unfinalised until we change the criteria of having voting power (and thereby definition of activated height)

My plan is to have this pr as the feature pr which collects all the sub-prs (defined in the TODO)

> I think if we do this, then BTC staking module is not repsonsible for inferring voting power of FPs anymore. Rather its sole functionality is to manage state FPs / BTC delegations. Voting power table then turns to be like a "view" of finality module over the BTC staking module.

Need to think about this more but seems this is out of scope of this feature? Better to have an RFC about this. Wdyt?

Reply:
>There is an ancient discussion on this https://babylonlabs.atlassian.net/wiki/spaces/BABYLON/pages/5878176/ADR-018+BTC+staking+consumer+module+refactoring?focusedCommentId=5878896 and it seems that the timestamping pub rand feature mandates this refactor

Oh I am big fan of such refactoring, to treat `btcstaking` module as a mostly dumb storage for fps and delegations. Though I agree that combining big security related feature and such refactoring is no go. 

>Would it cause any concerns if we let the btc staking keeper have some expected interface of finality keeper to this end?

Ultimately it may work, but we already have finality keeper depending on btc staking keeper, so if we would introduce dependency of finality keeper on btc staking keeper it would create circular dependency. Such cases usually mean there is some design problem and can lead to subtle bugs, for example, one needs to remember which dependency must be initialised first. 

Ideally we would end up with some implementation without such circular dependency and make refactoring later, though I wonder whether it is possible.

Reply:
I don't think circular dependency is an issue because we have used `expected_interfaces` to avoid go module dependency, i.e., the finality module does not directly depend on the btc staking module.

Reply:
>I don't think circular dependency is an issue because we have used expected_interfaces to avoid go module dependency, i.e., the finality module does not directly depend on the btc staking module.

Tbh I do not have in mind exact golang module dependency, but rather design dependency. If we now start to retrieve data from finality module in staking module, then it means:
- finality module relies on data from staking module
- staking module relies of data from finality module

This is circular and with time it can lead to some problems.

Reply:
In fact, this circular dependency will be catched by compiler -- the constructor functions of two keepers will depend the other keeper and `app.go` will not compile or panick due to nil pointer. Thus if we want to make activation height depend on pub rand, then moving voting power table to finality module is necessary.

So given that the timestamping of pub rand involves multiple steps, how about we do the following:

- create a feature branch for timestamping of pub rand merge this PR to the feature branch
- do a PR for moving voting power table to finality module, and this PR goes to `dev` branch (the integration team will merge this PR to the base branch as well)
- rebase the feature branch
- do a PR for excluding FPs without pub rand from voting power table at each height, and this PR goes to the feature branch

Reply:
Good point. So, we need to do three things in sequence:
1. introduce committed epoch (this PR does)
2. move the voting power table to the finality module
3. exclude fps without timestamped pub rand from the active set

I think all of the three tasks are independent and can be merged to `dev` in sequence. Another good thing is that after (1) is merged to `dev`, the finality provider side can be implemented in parallel. Wdty?

Reply:
Let's do what you proposed then just in case we have some important add-on to deploy before the whole timestamping feature is ready

Reply:
> In fact, this circular dependency will be catched by compiler -- the constructor functions of two keepers will depend the other keeper and app.go will not compile or panick due to nil pointer. Thus if we want to make activation height depend on pub rand, then moving voting power table to finality module is necessary.

I think you can get around it, by not using constructor, by just setting one keeper through pointer:
```
app.BTCStakingKepper.FinalityKepper = fk
```
it is ugly but it probably works. Though, I agree lets do it in proper way. 

>create a feature branch for timestamping of pub rand, and merge this PR to the feature branch
do a PR for moving voting power table to finality module, and this PR goes to dev branch (the integration team will merge this PR to the base branch as well)
rebase the feature branch
do a PR for excluding FPs without pub rand from voting power table at each height, and this PR goes to the feature branch

This sounds good, as Babylon is always ready to be released with the current set of auxiliary programs on devnet.

Reply:
One thing to remember,  we need to rememver that moving voting table from btcstaking module to finality module is also an breaking change in which in theory, we need to have upgrade handler that moves all data from `btcstaking` key space to `finality` key space. 

Of course this is only necessary if we need to do the migration even if database is empty.